### PR TITLE
Refactored pattern  many_primary_ctors

### PR DESCRIPTION
**Pull request by Vitaly-Protasov**
Solved in case of #528

### CODE REVIEW

File: aibolit/patterns/many_primary_ctors/many_primary_ctors.py [hunk 1]
```diff
@@ -19,42 +19,56 @@
...
+        lines: List[int] = list()
+        ast = AST.build_from_javalang(build_ast(filename))
+        for class_declaration in ast.get_proxy_nodes(ASTNodeType.CLASS_DECLARATION):
+            primary_lines = self.__find_primary(ast, class_declaration.body)
+            if len(primary_lines) > 1:
+                lines.extend(primary_lines)
 
         return lines
 
+    def __find_primary(self, ast: AST, class_body: List[ASTNode]) -> List[int]:
```
Comment (aravij) [inline aibolit/patterns/many_primary_ctors/many_primary_ctors.py:66]: We better stick to previous decomposition:
 - iterate over all constructors
 - check single constructor is it primary or not.

File: aibolit/patterns/many_primary_ctors/many_primary_ctors.py [hunk 2]
```diff
@@ -19,42 +19,56 @@
...
-    def __traverse_node(self, tree: javalang.ast.Node) -> List[LineNumber]:
-        lines: List[LineNumber] = list()
-
-        for _, class_declaration in tree.filter(javalang.tree.ClassDeclaration):
-            primary_ctors = list(filter(_is_primary, class_declaration.constructors))
-
-            if len(primary_ctors) > 1:
-                lines.extend(ctor.position.line for ctor in primary_ctors)
+    '''
+    If there is more than one primary
```
Comment (aravij) [inline aibolit/patterns/many_primary_ctors/many_primary_ctors.py:52]: Primary and secondary constructors need better clarification. May be an example from Yegor blog (https://www.yegor256.com/2015/05/28/one-primary-constructor.html)

File: aibolit/patterns/many_primary_ctors/many_primary_ctors.py [hunk 3]
```diff
@@ -19,42 +19,56 @@
...
+            if len(primary_lines) > 1:
+                lines.extend(primary_lines)
 
         return lines
 
+    def __find_primary(self, ast: AST, class_body: List[ASTNode]) -> List[int]:
+        lines: List[int] = []
+        for node in class_body:
+            if node.node_type == ASTNodeType.CONSTRUCTOR_DECLARATION:
+                for assignment in ast.get_subtree(node).get_proxy_nodes(ASTNodeType.ASSIGNMENT):
```
Comment (aravij) [inline aibolit/patterns/many_primary_ctors/many_primary_ctors.py:70]: According to Yegor post, I think it is easier to detect secondary constructors. They consist only of a single statement. All other constructors are primary.

Comment (lyriccoder) [reply aibolit/patterns/many_primary_ctors/many_primary_ctors.py]: at the moment private constructors are implemented correctly. Constructor is private when we have at least one variable assignment. Secondary constructor is constructor  when we do not have variable assignments and we have a call of primary constructor

**Comment by acheshkov** [comment]
@rultor merge

**Comment by acheshkov** [comment]
@rultor merge

**Pull request merged**